src/subtype.c

```diff
@@ -2535,6 +2535,31 @@ static int var_occurs_inside(jl_value_t *v, jl_tvar_t *var, int inside, int want
     return 0;
 }
 
+static jl_value_t *omit_bad_union(jl_value_t *u, jl_tvar_t *t) JL_NOTSAFEPOINT
+{
+    if (jl_is_unionall(u) && ((jl_unionall_t *)u)->var != t) {
+        if (jl_has_typevar(((jl_unionall_t *)u)->var->lb, t))
+            return NULL;
+        jl_value_t *ub = omit_bad_union(((jl_unionall_t *)u)->var->ub, t);
+        jl_value_t *body = omit_bad_union(((jl_unionall_t *)u)->body, t);
+        if (body == NULL || ub == NULL)
+            return NULL;
+        ((jl_unionall_t *)u)->var->ub = ub;
+        ((jl_unionall_t *)u)->body = body;
```

Comment:
Directly corrupting immutable/constant memory is not going to be safe. You need to allocate and return a new object.